### ISSUE TITLE
Rename smt_elevate_account -> smt_create

### ISSUE THREAD

**Issue by theoreticalbts** [steemit/steem#1554]
This operation needs to be renamed.

**Comment by theoreticalbts** [comment]
In addition to renaming I'm going to add the field for the NAI and serialization code to put it in an asset identifier.

**Comment by vogel76** [comment]
Following doubts must be clarified:
1) Does single control account can create multiple SMTs (having different NAIs) ? This is reasonable according to separation from controlled account and SMT as it was proposed initially.
2) When given NAI shall be reserved and stored in `smt_creation_operation` ? Operation evaluation stage is reasonable, especially the spec describes that NAI is allocated by blockchain.
3) How SMT symbol uniqueness shall be determined ? SMT ids share global space. `asset_symbol_type::validate` actually checks correctness of given SMT symbol according to NAI/precision encoding rules. Which rules shall be used to determine SMT uniquenes:
a) whole encoded asset_symbol_type is used to check it (so pair NAI, precision determines uniqueness)
b) part of encoded asset_symbol_type holding a NAI is only used. This disallows to create 2 SMTs having same NAI and different precision

**Comment by mvandeberg** [comment]
1. A single account can create multiple SMTs each having a different NAI.

2. The NAI is stored in `smt_creation_operation` when the user creates the application. In this case, we assume the user is some application providing a GUI to an actual human. The NAI is retrieved through an API call, `get_next_smt_identifier`.

`get_next_smt_identifier` takes a single argument, the desired precision of the SMT and returns a list of available NAIs. One of those NAIs is chosen and used in the operation. In #1633 I proposed creating a method called, `claim_smt_identifier` that takes a reference to the database and the NAI. The method will return true on a successful claim and false otherwise. Assert the return is true. This only updates the list of potential NAIs, but it ensures that the requested NAI will not be used by another SMT. The evaluator still needs to create the SMT object with the correct identifier.

For purposes of testing, define a valid NAI and have `claim_smt_identifier` return true only on the valid NAI. That way you can create valid and invalid ops and have the helper method respond accordingly. It will be implemented in a different issue.

3. The helper methods that determine the pool of potential NAIs will handle NAI uniqueness. It is not something that this evaluator needs to be concerned with. If `claim_smt_identifier` returns true then the NAI was already determined to be unique. 

The NAI without the precision digit will be unique. The precision digit only exists to represent the entirety of the symbol.

**Comment by Kiwonik** [comment]
With the latest commit both 1 (A single account can create multiple SMTs each having a different NAI.) and 3 (The NAI without the precision digit will be unique) are implemented.
As for 2 (NAI reservation) I understand that NAI reservation interface is not created yet, hence no place to put even the temporary testing variant of the `claim_smt_identifier`. Therefore now the NAI is aquired using method dedicated for the purpose `name_to_asset_symbol` as described by @theoreticalbts

**Comment by vogel76** [comment]
This point of implementation allows to continue (after accepted PR) other tasks basing on this one. But to continue smt_creation process, we definitely need to know:
- localisation and prototype of 'claim_smt_next_identifier' to continue work at evaluation stage
- localisation and prototype of 'get_smt_next_identifier' to continue work on unit testing scenarios

**Comment by mvandeberg** [comment]
@theoreticalbts Will prototype the methods so you can continue development.

**Comment by Kiwonik** [comment]
Waiting for promised prototypes to continue.

**Comment by youkaicountry** [comment]
A new issue has been created and assigned to @theoreticalbts for implementation of the claim_smt_identifier prototype method.

**Comment by youkaicountry** [comment]
For now, let's have smt_create accept any NAI for testing. This will allow us to move forward, and when the NAI claiming functionality is working we can hook it up at that point.

**Comment by Kiwonik** [comment]
No, accepting any NAI won't work, not even for testing. The reason is that NAI validation assertions have already been introduced in several places (and rightfully so), see e.g. issue #1738
Here's the idea to move forward with SMT creation without breaking existing code - Let us implement dummy (but correct!) versions of `claim_smt_next_identifier` and `get_smt_next_identifier`. All we need is a decision on localisation and prototype (interface) of these two method. It won't take more than 5 minutes of @theoretical 's precious time.

**Comment by Theoretical** [comment]
hey uhhh, you tagged the wrong guy  here >:(

**Comment by youkaicountry** [comment]
@theoreticalbts ^

**Comment by theoreticalbts** [comment]
> NAI validation assertions have already been introduced in several places

What assertions in what places?

> No, accepting any NAI won't work, not even for testing.

What specifically does the client have to do to generate an NAI that will be accepted?  If it's not too hard, maybe we should just add API support so clients can do that thing?

**Comment by vogel76** [comment]
@theoreticalbts Look at `asset_symbol_type::asset_num_from_nai` and `void asset_symbol_type::validate()const` where are located checks validating NAI. Operations creating SMTs are calling validate, what is fine.
We don't want to change (actually break) again code being already done, so better is to make such fake implementations, which can serve us some fake NAIs, able to satisfy checks. But to do so, we need to know at least prototypes and location of such functions.

**Comment by theoreticalbts** [comment]
As specified in the comments in `asset_symbol.hpp`, there are two different possible binary serializations of `asset_symbol_type`.  The *legacy binary serialization* accepts exactly three different 8 byte patterns:

```
TBD   03 54 42 44 00 00 00 00
TESTS 03 54 45 53 54 53 00 00
VESTS 06 56 45 53 54 53 00 00
```

SMT assets will use the *NAI serialization*.  The NAI serialization has the low 4 bits of the first byte represent the decimal place, identical to the legacy binary serialization.  The 5th bit of the first byte is `1`.  The remaining 27 bits can be anything, except the integer value represented by these 27 bits must be between `1` and `99999999` (as defined by `SMT_MIN_NAI` and `SMT_MAX_NAI`).

Right now, the code for SMT creation should allow literally any 4-byte sequence that satisfies these restrictions.  Writing such code is trivial because the binary serialization / deserialization code that implements these restrictions already exists in `pack()` / `unpack()` overrides in `asset_symbol.hpp`.  You simply declare a field of `asset_symbol_type` or `asset` type, and list the field in the `FC_REFLECT()` declaration.

Like any method named `validate()` should do, the `asset_symbol_type::validate()` and `asset::validate()` methods *only do stateless checks for internal consistency of the data structure*.

Now *completely separate from the stateless validate() check described above,* eventually there will be a *stateful restriction on the NAI's which can be created at any given time*.  The details of how this stateful restriction works are not important; don't worry about the stateful check for now.  It is not necessary for the initial testing version.  It will be done later.

**Comment by vogel76** [comment]
@theoreticalbts @youkaicountry Thanks for explanation how to generate valid NAI, this for sure will help us. But we definitely need to know prototypes and location of these 2 methods (`claim_smt_next_identifier` and `get_smt_next_identifier`) to be able use them in code. Few months (!) ago you have discusses this code to Vandeberg, so maybe finally provide us this basic information.

**Comment by theoreticalbts** [comment]
OK.  Eventually, when we implement the stateful restriction described in my previous comment, the client will need to know how to create an SMT identifier that passes the stateful restriction.  So we will give the client an RPC method to query the blockchain to understand what SMT identifiers currently pass the stateful restriction.  The name of this query method is `get_next_smt_identifier`.

So `get_next_smt_identifier` is only relevant when the stateful restriction is implemented.  It simply returns an NAI (or a set of NAI's) that has not been used and currently passes the stateful restriction.

There are some issues with `claim_next_smt_identifier` but I will discuss them in #1777.

**Comment by Kiwonik** [comment]
@youkaicountry As the latest commit included the temporary dummy implementation of `get_next_smt_identifier`, which is now used in unit tests I find this issue complete. Remaining work related to the method or smt creation has been described in separate issues ( #1729 #1786 #1633 ) so this issue may be closed once PR 1804 is merged.

**Comment by youkaicountry** [comment]
@Kiwonik Great, please move this issue from "In Work" to "Review" on the project board.
